crypto/asn1/ameth_lib.c

```diff
@@ -277,7 +277,7 @@ void EVP_PKEY_asn1_free(EVP_PKEY_ASN1_METHOD *ameth)
 
 void EVP_PKEY_asn1_set_public(EVP_PKEY_ASN1_METHOD *ameth,
                               int (*pub_decode) (EVP_PKEY *pk,
-                                                 X509_PUBKEY *pub),
+                                                 const X509_PUBKEY *pub),
```

Comment:
Do we consider this an API breaking change?

External applications will be required to change their source code, see this failure: https://travis-ci.org/github/romen/openssl/jobs/691216006

I am unsure if the fix is to update the gost engine to use the constified version (but this would also mean that sharing the codebase with 1.1.1 and 3.0.0-dev, this would require `#ifdef` guards) or to undo the constification.

Reply:
Ping @mattcaswell @t8m

Reply:
Current gost-engine upstream is not compatible with 1.1.1 already

Reply:
> Current gost-engine upstream is not compatible with 1.1.1 already

Ok, so this particular failure could be fixed by updating the gost source.

Yet, I want to raise the issue that any external app or engine that was using `EVP_PKEY_asn1_set_public()` will require a source change after the constification: it's not like we don't allow this for 3.0.0, but usually we have a conversation (if not a vote) about breaking API changes, and I don't think the matter was raised here.

It is perfectly acceptable to say that we will stick with the API change, I just want it to be acknowledged and approved explicitly.

Reply:
I agree with you.

Reply:
Actually I don't understand what renaming X509_PUBKEY_cmp to X509_PUBKEY_eq
has to do at all with constification of pub_decode functions ?

Reply:
This was needed  to allow 
`int X509_PUBKEY_eq(const X509_PUBKEY *a, const X509_PUBKEY *b)`
rather than
`int X509_PUBKEY_eq(X509_PUBKEY *a, X509_PUBKEY *b)`
and also independently of that function the constification should make much sense.

Reply:
It's really a pity that for so many OpenSSL API functions `const` types for read-only parameters were not done right from the beginning.

Reply:
> This was needed to allow
> `int X509_PUBKEY_eq(const X509_PUBKEY *a, const X509_PUBKEY *b)`
> rather than
> `int X509_PUBKEY_eq(X509_PUBKEY *a, X509_PUBKEY *b)`
> and also independently of that function the constification should make much sense.

I understand, and often constifying the arguments of a function do not cause source code compatibility issues!
But in this case, by changing the argument of a function pointer declaration we end up with a breaking API change.

I don't disagree with constification, and I wouldn't vote against the breaking change in this case, and I share the sentiment that it would have been better if it was const to begin with!

I just want to raise the issue that, as happened with the gost source, external applications will get failures out of this change and normally we require the reviewers to acknowledge and approve this, or even call a vote in some cases! 

> It's really a pity that for so many OpenSSL API functions `const` types for read-only parameters were not done right from the beginning.

Agreed!

Reply:
How to deal with the API incompatibility here? Should there a voting?
Fortunately `EVP_PKEY_asn1_set_public()` is an 'expert-only' function.
Shall I document that API-breaking change, and how best to do that?

Reply:
I just found that I missed to add the `const` in a further line of `EVP_PKEY_ASN1_METHOD.pod`:
```
 int (*pub_decode) (EVP_PKEY *pk, X509_PUBKEY *pub);
```

Reply:
BTW strange that for `priv_decode()` and `param_decode()` the input parameter was already `const` but not for `pub_decode()`.

At the latest when documenting those functions in commit 751148e2b0d1048522d754d55319166435cdda86
the `const` issue introduced in commit 0f113f3ee4d should have been rectified.

Reply:
Or could we just cast-away the const when the callback is called? I know, ugly and should be documented at least.

Reply:
Again my question: How shall we proceed now with this (in my view minor) API incompatibility?
Is this something that we should decide/fix urgently before the release?

Reply:
BTW, a week ago I realized that @mattcaswell had brought up a very similar issue 11  months ago, which is still open, too: https://github.com/openssl/openssl/issues/9296